tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreSession.java

```diff
@@ -0,0 +1,241 @@
+   * @return An entity containing the metadata.
+   */
+  @VisibleForTesting
+  Entity saveMetadataToEntity(Key sessionKey) {
+    Entity.Builder sessionEntity = Entity.newBuilder(sessionKey)
+        .set(SessionMetadata.CREATION_TIME, getCreationTime())
+        .set(SessionMetadata.LAST_ACCESSED_TIME, getLastAccessedTime())
+        .set(SessionMetadata.MAX_INACTIVE_INTERVAL, getMaxInactiveInterval())
+        .set(SessionMetadata.IS_NEW, isNew())
+        .set(SessionMetadata.IS_VALID, isValid())
+        .set(SessionMetadata.THIS_ACCESSED_TIME, getThisAccessedTime());
+
+    // A negative time indicates that the session should never time out
+    if (getMaxInactiveInterval() >= 0) {
+      sessionEntity.set(SessionMetadata.EXPIRATION_TIME,
+          getLastAccessedTime() + getMaxInactiveInterval() * 1000);
+    }
+
+    return sessionEntity.build();
+  }
+
+  /**
+   * Serialize the session attributes into entities.
+   * @param attributeKeyFactory The key builder for the entities.
+   * @return A list of entities where the key correspond to the name of the attribute
+             and the property `value` to the serialized attribute.
+   * @throws IOException If an error occur during the serialization.
+   */
+  @VisibleForTesting
+  List<Entity> saveAttributesToEntity(KeyFactory attributeKeyFactory) throws
+      IOException {
+    Stream<Entity> entities = Collections.list(getAttributeNames()).stream()
+        .filter(name -> accessedAttributes.contains(name))
+        .filter(name -> isAttributeDistributable(name, getAttribute(name)))
+        .map(name -> serializeAttribute(attributeKeyFactory, name));
+
+    try {
+      return entities.collect(Collectors.toList());
+    } catch (UncheckedIOException e) {
+      throw e.getCause();
+    }
+  }
+
+  /**
+   * Serialize an attribute an embed it into an entity whose key is generated by the provided
+   * KeyFactory.
+   * @param attributeKeyFactory The KeyFactory to use to create the key for the entity.
+   * @param name The name of the attribute to serialize.
+   * @return An Entity containing the serialized attribute.
+   */
+  private Entity serializeAttribute(KeyFactory attributeKeyFactory, String name) {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    try (ObjectOutputStream oos = new ObjectOutputStream(bos)) {
+      oos.writeObject(getAttribute(name));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    return Entity.newBuilder(attributeKeyFactory.newKey(name))
+        .set(SessionMetadata.ATTRIBUTE_VALUE_NAME,
+            BlobValue.newBuilder(Blob.copyFrom(bos.toByteArray()))
+                .setExcludeFromIndexes(true)
+                .build())
+        .build();
+  }
+
+  /**
+   * List the attributes that were loaded from the Datastore and suppressed during the execution.
+   * @return A set of the suppressed attributes.
+   */
+  public Set<String> getSuppressedAttributes() {
```

Comment:
I don't understand this logic - why are we suppressing attributes? What does it mean that they were loaded from the Datastore but then suppressed during execution? Please make it clearer in the comments.

Reply:
And why does it need to be public?

Reply:
Another suggestion we talked about: splitting out of the DatastoreSession class to Marshaller and Unmarshaller classes. This way the DatastoreSession would become a sort of PersistentSession whose responsible only to manage persistency related optimizations, that could be reused with other backends as well (e.g. Memcache, redis, etc.).

Reply:
Another suggestion we talked about: splitting out of the DatastoreSession class to Marshaller and Unmarshaller classes. This way the DatastoreSession would become a sort of PersistentSession whose responsible only to manage persistency related optimizations, that could be reused with other backends as well (e.g. Memcache, redis, etc.).

Reply:
Let's do the refactoring of the suppressed -> removed in another PR too.